php/Exchange.php

```diff
@@ -1552,47 +1552,40 @@ public function fetch($url, $method = 'GET', $headers = null, $body = null) {
         curl_setopt($this->curl, CURLOPT_FAILONERROR, false);
 
-        $response_headers = array();
-        $http_status_text = '';
-
-        // this function is called by curl for each header received
-        curl_setopt($this->curl, CURLOPT_HEADERFUNCTION,
-            function ($curl, $header) use (&$response_headers, &$http_status_text) {
-                $length = strlen($header);
-                $tuple = explode(':', $header, 2);
-                if (count($tuple) !== 2) { // ignore invalid headers
-                    // if it's a "GET https://example.com/path 200 OK" line
-                    // try to parse the "OK" HTTP status string
-                    if (substr($header, 0, 4) === 'HTTP') {
-                        $parts = explode(' ', $header);
-                        if (count($parts) === 3) {
-                            $http_status_text = trim($parts[2]);
-                        }
-                    }
-                    return $length;
-                }
-                $key = trim($tuple[0]);
-                $key = implode('-', array_map(get_called_class() . '::capitalize', explode('-', $key)));
-                $value = trim($tuple[1]);
-                if (!array_key_exists($key, $response_headers)) {
-                    $response_headers[$key] = $value;
-                } else {
-                    if (is_array($response_headers[$key])) {
-                        $response_headers[$key][] = $value;
-                    } else {
-                        $response_headers[$key] = array($response_headers[$key], $value);
-                    }
-                }
-                return $length;
-            }
-        );
+        curl_setopt($this->curl, CURLOPT_HEADER, 1);
+        // match the same http version as python and js
+        curl_setopt($this->curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
 
         // user-defined cURL options (if any)
         if (!empty($this->curl_options)) {
             curl_setopt_array($this->curl, $this->curl_options);
         }
 
-        $result = curl_exec($this->curl);
+        $response_headers = array();
+
+        $response = curl_exec($this->curl);
+
+        $headers_length = curl_getinfo($this->curl, CURLINFO_HEADER_SIZE);
+
+        $raw_headers = mb_substr($response, 0, $headers_length);
+
+        $raw_headers_array = explode("\r\n", trim($raw_headers));
+        $status_line = $raw_headers_array[0];
+        $parts = explode(' ', $status_line);
+        $http_status_text = count($parts) === 3 ? $parts[2] : null;
+        $raw_headers = array_slice($raw_headers_array, 1);
+        foreach ($raw_headers as $raw_header) {
+            list($key, $value) = explode(': ', $raw_header);
+            // don't overwrite headers
+            // https://stackoverflow.com/a/4371395/4802441
+            if (array_key_exists($key, $response_headers)) {
+                $response_headers[$key] = $response_headers[$key] . ', ' . $value;
```

Comment:
This will break the backward-compatibility, i'll need to rewrite this to match the old format of headers.